pkg/platform/api/buildlogstream/streamer.go

```diff
@@ -0,0 +1,155 @@
+package buildlogstream
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/gorilla/websocket"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/pkg/platform/api"
+	"github.com/ActiveState/cli/pkg/platform/model"
+
+	"github.com/sacOO7/gowebsocket"
+)
+
+type Request struct {
+	socket   gowebsocket.Socket
+	recipeID strfmt.UUID
+}
+
+func NewRequest(recipeID strfmt.UUID) *Request {
+	url := api.GetServiceURL(api.BuildLogStreamer)
+	socket := gowebsocket.New(url.String())
+	socket.WebsocketDialer = websocket.DefaultDialer
+	socket.RequestHeader.Set("Origin", "https://"+url.Host)
+
+	logging.Debug("websocket created for %s (origin: %s)", socket.Url, socket.RequestHeader.Get("Origin"))
+
+	request := &Request{socket: socket, recipeID: recipeID}
+
+	return request
+}
+
+func (r *Request) close() {
+	r.socket.Close()
+}
+
+type logRequest struct {
+	RecipeID string `json:"recipeID"`
+}
+
+type artifactMapping struct {
+	name    *string
+	version *string
+}
+
+type message struct {
+	Type         string      `json:"type"`
+	CacheHit     bool        `json:"cache_hit"`
+	ErrorMessage string      `json:"error_message"`
+	ArtifactID   strfmt.UUID `json:"artifact_id"`
+}
+
+func (r *Request) Wait() error {
+	request := logRequest{RecipeID: r.recipeID.String()}
+
+	out := output.Get()
+
+	// Creative error handling, cause why would you just return the error right?
+	var err error
+	r.socket.OnConnectError = func(connectErr error, _ gowebsocket.Socket) {
+		logging.Debug("Connection error: %s", err)
+		err = connectErr
+	}
+
+	logging.Debug("connecting to head-chef websocket")
+	r.socket.Connect()
+	if err != nil {
+		return errs.Wrap(err, "Could not connect to websocket")
+	}
+
+	time.Sleep(time.Second)
```

Comment:
Sleeps like this are a bright redish-orange flag for me. Martin's concern about not using OnConnect seem quite relevant; The control flow is not as direct as it appears.

Without spending more time, I'm not exactly sure what `gowebsocket` is really offering us here since it's using `gorilla/websocket` underneath. Nevertheless, I think that layer is hiding the control flow and making it a bit more difficult to rationalize the control flow.

Reply:
This is WIP code, not ready for any sort of review. The sleep was to test a race condition.

Reply:
Understood... still, not using OnConnect and SendBinary/SendText would mean that it's possible to be interacting with connection in a poor state.